rmw_dds_common/test/test_graph_cache.cpp

```diff
@@ -0,0 +1,1113 @@
+    // topic4 writers
+    {"writer8", "topic4", "Int", false},
+  });
+
+  // Check graph state.
+  check_results(
+    graph_cache,
+    {},
+  {
+    {"topic1", {"Str"}},
+    {"topic2", {"Bool", "Float", "Int", "Str"}},
+    {"topic3", {"Float"}},
+    {"topic4", {"Int"}},
+  });
+  check_results_by_node(graph_cache, "ns", "name");
+  check_results_by_topic(graph_cache, "topic1", 4, 2);
+  check_results_by_topic(graph_cache, "topic2", 2, 3);
+  check_results_by_topic(graph_cache, "topic3", 1, 0);
+  check_results_by_topic(graph_cache, "topic4", 0, 1);
+
+  // Remove some readers and writers.
+  remove_entities(
+    graph_cache,
+  {
+    // topic1
+    {"reader2", "topic1", "Str", true},
+    {"reader3", "topic1", "Str", true},
+    {"reader4", "topic1", "Str", true},
+    {"writer2", "topic1", "Str", false},
+    // topic2
+    {"reader6", "topic2", "Int", true},
+    {"writer5", "topic2", "Str", false},
+    {"writer6", "topic2", "Float", false},
+    {"writer7", "topic2", "Bool", false},
+    // topic3
+    {"reader7", "topic3", "Float", true},
+  });
+
+  // Check graph state.
+  check_results(
+    graph_cache,
+    {},
+  {
+    {"topic1", {"Str"}},
+    {"topic2", {"Str"}},
+    {"topic4", {"Int"}},
+  });
+  check_results_by_node(graph_cache, "ns", "name");
+  check_results_by_topic(graph_cache, "topic1", 1, 1);
+  check_results_by_topic(graph_cache, "topic2", 1, 0);
+  check_results_by_topic(graph_cache, "topic3", 0, 0);
+  check_results_by_topic(graph_cache, "topic4", 0, 1);
+
+  // Remove the remaining readers and writers.
+  remove_entities(
+    graph_cache,
+  {
+    // topic1
+    {"reader1", "topic1", "Str", true},
+    {"writer1", "topic1", "Str", false},
+    // topic2
+    {"reader5", "topic2", "Str", true},
+    // topic4
+    {"writer8", "topic4", "Int", false},
+  });
+
+  // Check graph state.
+  check_results(graph_cache);
+  check_results_by_node(graph_cache, "ns", "name");
+  check_results_by_topic(graph_cache, "topic1");
+  check_results_by_topic(graph_cache, "topic2");
+  check_results_by_topic(graph_cache, "topic3");
+  check_results_by_topic(graph_cache, "topic4");
+}
+
+void
+add_participants(
+  GraphCache & graph_cache,
+  const std::vector<std::string> & gids)
+{
+  for (const auto & gid : gids) {
+    graph_cache.add_participant(gid_from_string(gid));
+  }
+}
+
+void
+remove_participants(
+  GraphCache & graph_cache,
+  const std::vector<std::string> & gids)
+{
+  for (const auto & gid : gids) {
+    graph_cache.remove_participant(gid_from_string(gid));
+  }
+}
+
+rmw_dds_common::msg::Gid
+gid_msg_from_string(const std::string & str)
+{
+  rmw_dds_common::msg::Gid gid;
+  gid.data = {0};
+  EXPECT_LT(str.size(), RMW_GID_STORAGE_SIZE);
+  std::strncpy(
+    reinterpret_cast<char *>(gid.data.data()),
+    str.c_str(),
+    RMW_GID_STORAGE_SIZE);
+  return gid;
+}
+
+struct NodeInfo
+{
+  std::string gid;
+  std::string namespace_;
+  std::string name;
+};
+
+using NodeInfoVec = std::vector<NodeInfo>;
```

Comment:
@ivanpauno nit: I fail to see the advantage of `NodeInfoVec` over `std::vector<NodeInfo>`.

Reply:
Yeah, it's just one Hamming distance away ....
Solved in https://github.com/ivanpauno/rmw_dds_common/pull/4/commits/3f8d980c1547482f78292da00ff41bc0c501f582.